src/token/api/src/storage/structures.rs

```diff
@@ -0,0 +1,102 @@
+use std::collections::{hash_map::Entry, HashMap};
+
+use candid::Principal;
+use canister_sdk::ic_kit::ic;
+use ic_stable_structures::{btreemap, cell, log, memory_manager::MemoryId, Storable};
+
+use super::Memory;
+
+pub struct StableCell<T: Storable> {
+    data: HashMap<Principal, cell::Cell<T, Memory>>,
+    default_value: T,
+    memory_id: MemoryId,
+}
+
+impl<T: Storable> StableCell<T> {
+    pub fn new(memory_id: MemoryId, default_value: T) -> Self {
+        Self {
+            data: HashMap::default(),
+            default_value,
+            memory_id,
+        }
+    }
+
+    pub fn get(&self) -> &T {
+        let canister_id = ic::id();
+        self.data
+            .get(&canister_id)
+            .map(|cell| cell.get())
+            .unwrap_or(&self.default_value)
+    }
+
+    /// Updates value in stable memory.
+    pub fn set(&mut self, value: T) {
+        let canister_id = ic::id();
+        match self.data.entry(canister_id) {
+            Entry::Occupied(mut entry) => {
+                entry
+                    .get_mut()
+                    .set(value)
+                    .expect("failed to set value to stable cell");
```

Comment:
Return an error, do not panic here. You don't know at what point this code is executed and if it's ok to panic there or not.

Reply:
Done. Now expects are in client code, which knows that the only reason for errors is out of stable memory, which we can't handle somehow.